cpp/src/arrow/dataset/file_base.cc

```diff
@@ -465,14 +465,7 @@ class TeeNode : public compute::MapNode {
       // Need to wait for the task group to complete regardless of dw_status
       task_group_.End().AddCallback(
           [this, dw_status, finish_st](const Status& tg_status) {
-            // Prefer dw_status then finish_st and then tg_status
-            if (!dw_status.ok()) {
-              finished_.MarkFinished(dw_status);
-            }
-            if (!finish_st.ok()) {
-              finished_.MarkFinished(finish_st);
-            }
-            finished_.MarkFinished(tg_status);
+            finished_.MarkFinished(dw_status & finish_st & tg_status);
```

Comment:
Nothing to change, just a side comment: I think it is non-standard that `Status::operator&` is short-circuiting, as this is normally the semantics of `operator&&`.

Reply:
Agreed.  I'm not sure why it was chosen to be this way.  Perhaps inspired by python where `&&` forces the return of the operator to be a boolean.